From 1eafb446656fe7c314066586f68d596273980073 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jo=C3=A3o=20Henrique=20Ferreira=20de=20Freitas?=
 <joaohf@gmail.com>
Date: Fri, 26 Feb 2021 17:11:35 -0300
Subject: [PATCH] Do not discard parameters --system_libs and --include-erts

If the parameters --system_libs and --include-erts from rebar3 tar or
release are passed, then they need to take into account when loading
relx configuration file.

This commits checkes if the parameters are passed. If yes, they will be
used to override any system_libs and/or include_erts.

Fixes 2496

Upstream-Status: Pending
---
 src/rebar_relx.erl | 24 +++++++++++++++++++++++-
 1 file changed, 23 insertions(+), 1 deletion(-)

diff --git a/src/rebar_relx.erl b/src/rebar_relx.erl
index af4104f3..8d48cc16 100644
--- a/src/rebar_relx.erl
+++ b/src/rebar_relx.erl
@@ -38,7 +38,10 @@ do(Provider, State) ->
                                {overlay_vars_values, ExtraOverlays},
                                {overlay_vars, [{base_dir, rebar_dir:base_dir(State)}]}
                                | merge_overlays(RelxConfig)],
-    {ok, RelxState} = rlx_config:to_state(RelxConfig1),
+
+    RelxConfig2 = maybe_include_erts_and_system_libs(RelxConfig1, Opts),
+
+    {ok, RelxState} = rlx_config:to_state(RelxConfig2),
 
     Providers = rebar_state:providers(State),
     Cwd = rebar_state:dir(State),
@@ -195,6 +198,25 @@ merge_overlays(Config) ->
     NewOverlay = lists:flatmap(fun({overlay, Overlay}) -> Overlay end, lists:reverse(Overlays)),
     [{overlay, NewOverlay} | Others].
 
+maybe_include_erts_and_system_libs(RelxConfig, Opts) ->
+    C0 = case proplists:is_defined(include_erts, Opts) of
+             false ->
+                 RelxConfig;
+             true ->
+                 V0 = proplists:get_value(include_erts, Opts),
+                 lists:keyreplace(include_erts, 1, RelxConfig, {include_erts, V0})
+         end,
+
+    C1 = case proplists:is_defined(system_libs, Opts) of
+             false ->
+                 C0;
+             true ->
+                 V1 = proplists:get_value(system_libs, Opts),
+                 lists:keyreplace(system_libs, 1, C0, {system_libs, V1})
+         end,
+
+    C1.
+
 %%
 
 %% Returns a map of all apps that are part of the rebar3 project.
-- 
2.25.1

